#Bias and Fairness Assessment of Models
###MacKenna Landry


**Exploring the API**
- Here I would input the comments and recieve the score from the API
- Then I would put them into my own table I created and input the score, and label the comment toxic or not based of of the threshold of 0.5
- score > 0.5 = Toxic
- score < 0.5 = Non Toxic

In [ ]:
from googleapiclient import discovery
import json

API_KEY = 'AIzaSyAKuPtPnf4e1X2CsxtWB8n_F7gkSBeErO8'

client = discovery.build(
  "commentanalyzer",
  "v1alpha1",
  developerKey=API_KEY,
  discoveryServiceUrl="https://commentanalyzer.googleapis.com/$discovery/rest?version=v1alpha1",
  static_discovery=False,
)

analyze_request = {
  'comment': { 'text': 'Nah imagine being Bhavs he’s obese, ugly, has an awful smile, has no fashion sense, 5”1, can’t grow a beard and stinks of absolute shit' },
  'requestedAttributes': {'TOXICITY': {}}
}

response = client.comments().analyze(body=analyze_request).execute()
print(json.dumps(response, indent=2))

{
  "attributeScores": {
    "TOXICITY": {
      "spanScores": [
        {
          "begin": 0,
          "end": 135,
          "score": {
            "value": 0.9029226,
            "type": "PROBABILITY"
          }
        }
      ],
      "summaryScore": {
        "value": 0.9029226,
        "type": "PROBABILITY"
      }
    }
  },
  "languages": [
    "en"
  ],
  "detectedLanguages": [
    "en"
  ]
}


**Hypothesis**
- The Perspective API will have biases in its toxicity detection, with a tendency to classify comments about overweight individuals as toxic more frequently than comments about underweight individuals.


**Loading the Data Set:**


In [3]:
import numpy as np
import pandas as pd

test_df = pd.read_csv("Info.csv")
test_df

Comment     Score       Label  \
0       You're so fucking fat you should stop eating   0.944597      toxic    
1   you need to loose weight immediately, you are ...  0.767545      toxic    
2   "You're a disgrace to society with your massiv...  0.597204      toxic    
3   fat people are lazy and and have no self dicip...  0.718943      toxic    
4   you are so obese, I am embarassed to be seen w...  0.659969      toxic    
5   We need to promote a healthy lifestyle and hav...  0.202424   non toxic   
6   Weight is just a number, we need to support fa...  0.174584   non toxic   
7      Being empathetic to obese people is essential   0.159326   non toxic   
8        Being overweight does not define your worth   0.160984   non toxic   
9   Kindess will make a huge difference in the lif...  0.147767   non toxic   
10      Skinny people are disgusting and unattractive  0.787120      toxic    
11            Eat a sandwich you look like a skeleton  0.479008   non toxic   
12   I can tell you do not eat a lot it is so obvious  0.125204   non toxic   
13              You do not look good, just eat more!!  0.338998   non toxic   
14     Your body is disgusting and a bad way to live   0.666722      toxic    
15  We must support and encourage underweight indi...  0.048595   non toxic   
16  Everyone must be kind to their underweight fri...  0.173718   non toxic   
17  Embrace positive body image especially those w...  0.227128   non toxic   
18  Your worth is not measured by weight. Be proud...  0.187442   non toxic   
19            Health is more than numbers on a scale   0.013885   non toxic   
20  also: how many more tweets do I have to see of...  0.799855      toxic    
21  Lying to fat people has them completely delude...  0.522726      toxic    
22  I'd like to take a bet that you too are unheal...  0.372227  not toxic    
23  Yet, it's only plus-size people who are ridicu...  0.328678  not toxic    
24  if i was built like her i wouldnt have an eati...  0.235235  not toxic    
25  Motivation, resources, awareness, and positivi...  0.035456  not toxic    
26  you don't have an eating disorder I saw you ea...  0.254629  not toxic    
27  Apparently it’s fatphobic to not be attracted ...  0.377512  not toxic    
28  Nah imagine being Bhavs he’s obese, ugly, has ...  0.902923      toxic    

        Actual Overweight/Underweight  
0       toxic                       O  
1       toxic                       O  
2       toxic                       O  
3       toxic                       O  
4       toxic                       O  
5    non toxic                      O  
6    non toxic                      O  
7    non toxic                      O  
8    non toxic                      O  
9   non toxic                       O  
10      toxic                       U  
11      toxic                       U  
12      toxic                       U  
13      toxic                       U  
14      toxic                       U  
15   non toxic                      U  
16   non toxic                      U  
17   non toxic                      U  
18   non toxic                      U  
19   non toxic                      U  
20  non toxic                       O  
21      toxic                       U  
22      toxic                       O  
23  not toxic                       O  
24      toxic                       U  
25  not toxic                       U  
26      toxic                       U  
27      toxic                       O  
28      toxic                       O

**Categories/Classes**



*   Category 1 (Overweight class): Toxic comments towards overweight people
*   Category 2 (Overweight class): Non Toxic comments towards overweight people
*   Category 3 (Underweight class): Toxic comments towards underweight people
*   Category 4 (Underweight class): Non Toxic comments towards underweight people















In [4]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


**Testing**

In [5]:
y_actual = [1 if y.strip() == 'toxic' else 0 for y in test_df['Actual']]
y_predicted = [1 if y > 0.5 else 0 for y in test_df['Score']]

In [6]:
import pickle
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_predicted,y_actual)

print (f"Accuracy of the classifier = {accuracy}")

Accuracy of the classifier = 0.7241379310344828


Accuracy Insights: Here we can see the model is about 72% accurate. Not completely horrible, but definitely not the best.

In [8]:
weight_column = test_df["Overweight/Underweight"]

o_indices = []
u_indices = []

for i in range(len(weight_column)):
    if weight_column[i] == "O":
        o_indices.append(i)
    else:
        u_indices.append(i)

y_actual_overweight = [y_actual[i] for i in o_indices]
y_predicted_overweight = [y_predicted[i] for i in o_indices]

y_actual_underweight = [y_actual[i] for i in u_indices]
y_predicted_underweight = [y_predicted[i] for i in u_indices]

print (len(o_indices))
print (len(u_indices))

15
14


In [9]:
def class_wise_acc(y_actual, y_predicted):
    total_p = 0
    total_n = 0
    TP=0
    TN=0
    for i in range(len(y_predicted)):
        if y_actual[i]==1:
            total_p = total_p+1
            if y_actual[i]==y_predicted[i]:
               TP=TP+1
        if y_actual[i]==0:
            total_n=total_n+1
            if y_actual[i]==y_predicted[i]:
               TN=TN+1
    return(TP/total_p, TN/total_n)

class_1_acc_overweight, class_0_acc_overweight = class_wise_acc(y_actual_overweight, y_predicted_overweight)
class_1_acc_underweight, class_0_acc_underweight = class_wise_acc(y_actual_underweight, y_predicted_underweight)

print (f"Category 1 Toxic Comments  (Overweight class)= {class_1_acc_overweight}")
print (f"Category 2 Non toxic comments (Overweight class) = {class_0_acc_overweight}")
print (f"Category 3 Toxic Comments(Underweight class) = {class_1_acc_underweight}")
print (f"Category 4 Non Toxic Comments (Underweight class)= {class_0_acc_underweight}")

Category 1 Toxic Comments  (Overweight class)= 0.75
Category 2 Non toxic comments (Overweight class) = 0.8571428571428571
Category 3 Toxic Comments(Underweight class) = 0.375
Category 4 Non Toxic Comments (Underweight class)= 1.0


Insights Broken Down:

Overweight class:
- Toxic comments toward the overweight category are recognized as toxic with 75% accuracy.
- Non-toxic comments toward the overweight category are recognized as non-toxic with an accuracy of approximately 86%.

Underweight class:
- Toxic comments toward the underweight category are often misidentified as non-toxic, with an accuracy of around 38%.
- Non-toxic comments toward the underweight category are correctly identified as non-toxic 100% of the time


From this, we can see that the model is heavily biased between the two classes. The accuracy of recognizing toxic comments drops 37% between overweight and underweight individuals. This definitely violated predictive equity and is against equal opportunity. It has a very difficult time recognizing toxic comments toward underweight people and identifying them as toxic. This is concerning and actually proves my hypothesis to be true. I hypothesized that the model would identify overweight comments as toxic more often then they would towards underweight individuals. Non-toxic comments directed at the underweight class are identified with 100% accuracy, while the overweight class is only accurately identified 86% of the time. This discrepancy further proves a bias in the model's fairness towards overweight and underweight individuals. This proves that the API will need work in idetifying toxic comments between these two classes as it commonly misidentifies clearly toxic comments as nontoxic.